Create rule S6949: Don't hardcode resource locations (#3783)
SonarSource/rspec: rules/S6949/azureresourcemanager/rule.adoc

> When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. Locations can either be added in a template directly or passed as a parameter. It's advisable to use a parameter to specify the location for resources.

Reviewer: Suggestion by grammarly:
```suggestion
When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. Locations can either be added in a template directly or passed as a parameter. It's advisable to use a parameter to specify the location of resources.
```

Reviewer: 2 times more in this document changing `for` -> `of`

Reviewer: There are some other recommendations by Grammarly, which I found help clear the text a bit. Please have a look at them (I pasted the text into an empty gDoc)

Reviewer 2: Totally forgot that I used to use the Grammarly extension for writing rules. Thanks, I'll look through it.

== Why is this an issue?

> When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. This can be done directly in the template or by passing parameters. However, hardcoding locations in the template can limit flexibility and potentially create deployment challenges, as it restricts users from choosing their preferred deployment location.

Reviewer: ```
When deploying an Azure Resource Manager template (ARM template), you must provide a location for each resource. This can be done directly in the template or by passing parameters. 
```
This is basically duplicated to the previous section

Reviewer 2: Yes. This is an idea I remember since the LaYC sprint: these sections can be read independently, so the short summary is basically a shorter version of "Why is this an issue". I'm not sure if it make sense to word them more differently

Reviewer: Ok for me. 
But my 2 cents:
There are too many people working independently on rspec. We are never going to achieve consistency...
I would like dedicated people only for rspec that are not us :D

It is therefore recommended to use a parameter to specify the location for resources, with the default value set to `resourceGroup().location`. This practice not only ensures consistency in resource allocation but also provides users of the template the flexibility to specify a location where they have the necessary permissions to deploy resources. This approach helps avoid hardcoding locations, which can lead to potential deployment issues and restrictions.

== How to fix it in ARM templates

> To fix this code smell in ARM templates, create a parameter for the location and set the default value to `resourceGroup().location`. Then, use the parameter to specify the location for resources.

Reviewer: ```suggestion
Create a parameter for the location and set the default value to `resourceGroup().location`. Then, use the parameter to specify the location of resources.
```

Reviewer: I don't like this prefix `To fix this code smell in ARM templates`. It's mostly redundant information and doesn't add value. I know it gets suggested by IslandAI everytime.

Reviewer: Mh, now I see that it differentiates between ARM templates and Bicep. I still think it's not necessary. But only my opinion here...

Reviewer 2: I agree that this is redundant. Even if there were differences in how to fix it sections, it would be clear from the context

=== Code examples

==== Noncompliant code example

[source,json,diff-id=11,diff-type=noncompliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2019-06-01",
      "name": "[parameters('storageAccountName')]",
      "location": "westus",
      "sku": {
        "name": "Standard_LRS"
      },
      "kind": "StorageV2",
      "properties": {}
    }
  ]
}
----

> ==== Compliant solution

Reviewer: What about adding this as a compliant solution?
```
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2019-06-01",
      "name": "[parameters('storageAccountName')]",
      "location": "[resourceGroup().location]",
      "sku": {
        "name": "Standard_LRS"
      },
      "kind": "StorageV2"
    }
  ]
}
```

Reviewer 2: I don't think that's the best example, because the idea is to allow users to set location during deployment with a sensible default value. ,I would prefer to show the entire recommendation in a compliant example.

[source,json,diff-id=11,diff-type=compliant]
----
{
  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "location": {
      "type": "string",
      "defaultValue": "[resourceGroup().location]",
      "metadata": {
        "description": "Location for all resources."
      }
    }
  },
  "resources": [
    {
      "type": "Microsoft.Storage/storageAccounts",
      "apiVersion": "2019-06-01",
      "name": "[parameters('storageAccountName')]",
      "location": "[parameters('location')]",
      "sku": {
        "name": "Standard_LRS"
      },
      "kind": "StorageV2",

>       "properties": {}

Reviewer: I would suggest removing these 3 properties to have a more minimal example. I don't see that they are strictly related to the code example or necessary. Maybe I'm wrong though.

Reviewer: Ok, sku and kind are required for storage Accounts. But properties is not :D

    }
  ]
}
----

== How to fix it in Bicep

To fix this code smell in Bicep, create a parameter for the location and set the default value to `resourceGroup().location`. Then, use the parameter to specify the location for resources.

=== Code examples

==== Noncompliant code example
[source,bicep,diff-id=2,diff-type=noncompliant]
----
resource storageAccount 'Microsoft.Storage/storageAccounts@2019-06-01' = {
  name: storageAccountName
  location: 'westus'
  sku: {
    name: 'Standard_LRS'
  }
  kind: 'StorageV2'
}
----

==== Compliant solution

[source,bicep,diff-id=2,diff-type=compliant]
----
param location string = resourceGroup().location

resource storageAccount 'Microsoft.Storage/storageAccounts@2019-06-01' = {
  name: storageAccountName
  location: location
  sku: {
    name: 'Standard_LRS'
  }

>   kind: 'StorageV2'

Reviewer: Same as above.

}
----

== Resources
=== Documentation
* https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/resource-location/[Azure Resource Manager templates: Resource location]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message
Replace this hardcoded location with a parameter.

> 

Reviewer: Please add `=== Highlighting` information
